File: Session_12_BranchAndBound/multi_language_implementations/knapsack_branch_bound.cpp
// Session 12: 0/1 Knapsack using Branch and Bound in C++

#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

struct Node {
    int level, profit, weight;
    double bound;
};

struct Item {
    int id, weight, value;
};

bool compareItems(const Item &a, const Item &b) {
    double r1 = (double)a.value / a.weight;
    double r2 = (double)b.value / b.weight;
    return r2 < r1; // descending order
}

double calculateBound(Node u, int n, int W, const std::vector<Item> &items) {
    if (u.weight >= W) return 0.0;
    double profit_bound = u.profit;
    int j = u.level + 1;
    int total_weight = u.weight;
    while (j < n && total_weight + items[j].weight <= W) {
        total_weight += items[j].weight;
        profit_bound += items[j].value;
        j++;
    }
    if (j < n) {
        profit_bound += (W - total_weight) * ((double)items[j].value / items[j].weight);
    }
    return profit_bound;
}

int solveKnapsackBB(int W, std::vector<Item> &items, int n) {
    std::sort(items.begin(), items.end(), compareItems);
    std::queue<Node> q;
    Node u, v;
    u.level = -1; u.profit = 0; u.weight = 0;
    u.bound = calculateBound(u, n, W, items);
    int maxProfit = 0;
    q.push(u);

    while (!q.empty()) {
        u = q.front();
        q.pop();

        if (u.level == -1) v.level = 0;
        else if (u.level == n - 1) continue;
        else v.level = u.level + 1;

        // branch with item
        v.weight = u.weight + items[v.level].weight;
        v.profit = u.profit + items[v.level].value;
        if (v.weight <= W && v.profit > maxProfit) maxProfit = v.profit;
        v.bound = calculateBound(v, n, W, items);
        if (v.bound > maxProfit) q.push(v);

        // branch without item
        v.weight = u.weight; v.profit = u.profit;
        v.bound = calculateBound(v, n, W, items);
        if (v.bound > maxProfit) q.push(v);
    }
    return maxProfit;
}

int main() {
    std::vector<Item> items = {{1, 2, 40}, {2, 5, 30}, {3, 10, 50}, {4, 5, 10}};
    std::cout << "Max Profit: " << solveKnapsackBB(10, items, 4) << "\n";
    return 0;
}

File: Session_12_BranchAndBound/multi_language_implementations/knapsack_branch_bound.py
# Session 12: 0/1 Knapsack using Branch and Bound in Python

from collections import deque

class Node:
    def __init__(self, level=-1, profit=0, weight=0, bound=0.0):
        self.level = level
        self.profit = profit
        self.weight = weight
        self.bound = bound

class Item:
    def __init__(self, item_id, weight, value):
        self.id = item_id
        self.weight = weight
        self.value = value

def calculate_bound(u, n, W, items):
    if u.weight >= W:
        return 0.0
        
    profit_bound = float(u.profit)
    j = u.level + 1
    total_weight = u.weight
    
    while j < n and total_weight + items[j].weight <= W:
        total_weight += items[j].weight
        profit_bound += items[j].value
        j += 1
        
    if j < n:
        profit_bound += (W - total_weight) * (float(items[j].value) / items[j].weight)
        
    return profit_bound

def solve_knapsack_bb(W, items, n):
    # Sort items based on value/weight ratio in descending order
    items.sort(key=lambda item: float(item.value) / item.weight, reverse=True)
    
    queue = deque()
    u = Node(level=-1, profit=0, weight=0)
    u.bound = calculate_bound(u, n, W, items)
    max_profit = 0
    queue.append(u)
    
    while queue:
        u = queue.popleft()
        
        if u.level == -1:
            v_level = 0
        elif u.level == n - 1:
            continue
        else:
            v_level = u.level + 1
            
        # Branch with item
        v_with = Node(
            level=v_level,
            profit=u.profit + items[v_level].value,
            weight=u.weight + items[v_level].weight
        )
        if v_with.weight <= W and v_with.profit > max_profit:
            max_profit = v_with.profit
            
        v_with.bound = calculate_bound(v_with, n, W, items)
        if v_with.bound > max_profit:
            queue.append(v_with)
            
        # Branch without item
        v_without = Node(
            level=v_level,
            profit=u.profit,
            weight=u.weight
        )
        v_without.bound = calculate_bound(v_without, n, W, items)
        if v_without.bound > max_profit:
            queue.append(v_without)
            
    return max_profit

def main():
    items = [
        Item(1, 2, 40),
        Item(2, 5, 30),
        Item(3, 10, 50),
        Item(4, 5, 10)
    ]
    print(f"Max Profit: {solve_knapsack_bb(10, items, 4)}")

if __name__ == "__main__":
    main()
